modules/default/weather/weatherprovider.js

```diff
@@ -111,36 +111,132 @@ const WeatherProvider = Class.extend({
 		this.delegate.updateAvailable(this);
 	},
 
-	getCorsUrl: function () {
-		if (this.config.mockData || typeof this.config.useCorsProxy === "undefined" || !this.config.useCorsProxy) {
-			return "";
-		} else {
-			return location.protocol + "//" + location.host + "/cors?url=";
-		}
-	},
-
 	/**
 	 * A convenience function to make requests.
 	 *
 	 * @param {string} url the url to fetch from
 	 * @param {string} type what contenttype to expect in the response, can be "json" or "xml"
+	 * @param {Array.<{name: string, value:string}>} requestHeaders the HTTP headers to send
+	 * @param {Array.<string>} expectedResponseHeaders the expected HTTP headers to recieve
 	 * @returns {Promise} resolved when the fetch is done
 	 */
-	fetchData: async function (url, type = "json") {
-		url = this.getCorsUrl() + url;
+	fetchData: async function (url, type = "json", requestHeaders = undefined, expectedResponseHeaders = undefined) {
+		url = this.getCorsUrl(url, requestHeaders, expectedResponseHeaders);
 		const mockData = this.config.mockData;
 		if (mockData) {
 			const data = mockData.substring(1, mockData.length - 1);
 			return JSON.parse(data);
 		} else {
 			const response = await fetch(url);
 			const data = await response.text();
+
 			if (type === "xml") {
 				return new DOMParser().parseFromString(data, "text/html");
 			} else {
-				return JSON.parse(data);
+				if (!data || !data.length > 0) return undefined;
+
+				const dataResponse = JSON.parse(data);
+				if (!dataResponse.headers) {
+					dataResponse.headers = this.getHeadersFromResponse(expectedResponseHeaders, response);
+				}
+				return dataResponse;
+			}
+		}
+	},
+
+	/**
+	 * Gets a URL that will be used when calling the CORS-method on the server.
```

Comment:
is this whole stuff necessary after you rmerge of the Support HTTP headers with CORS-method (#2957) ?

Reply:
Is this something that can be generalized so that other modules could use it too on the client side?

Reply:
Yes, I moved it to a `utils`-file now. This should make it easy to utilize for other modules.